src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/PropertyAccessor.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Reflection.Emit;
+
+namespace Raven.Server.Documents.Indexes.Persistence.Lucene.Documents
+{
+    public delegate object DynamicGetter(object target);
+
+    public class PropertyAccessor
+    {
+        public readonly Dictionary<string, DynamicGetter> Properties = new Dictionary<string, DynamicGetter>();
+
+        public static PropertyAccessor Create(Type type)
+        {
+            var accessor = new PropertyAccessor();
+
+            foreach (var prop in type.GetProperties())
+            {
+                accessor.Properties.Add(prop.Name, CreateGetMethod(prop, type));
+            }
+
+            return accessor;
+        }
+
+        public object GetValue(string name, object target)
+        {
+            DynamicGetter getterMethod;
+            if (Properties.TryGetValue(name, out getterMethod))
+                return getterMethod(target);
+
+            throw new InvalidOperationException(string.Format("The {0} property was not found", name));
+        }
+
+        private static DynamicGetter CreateGetMethod(PropertyInfo propertyInfo, Type type)
```

Comment:
I wonder if we should use expression trees here, to avoid having the reflection emit code here.
It would save a dependency, and it is more easy to read / work with.